cpu/sam0_common/periph/usbdev.c

```diff
@@ -0,0 +1,768 @@
+    _enable_irq(usbdev);
+
+    _block_pm();
+    usbdev->usbdev.cb(&usbdev->usbdev, USBDEV_EVENT_HOST_CONNECT);
+    /* Interrupt configuration */
+    NVIC_EnableIRQ(USB_IRQn);
+}
+
+static void usb_attach(sam0_common_usb_t *dev)
+{
+    DEBUG("sam_usb: Attaching to host\n");
+    dev->config->device->CTRLB.reg &= ~USB_DEVICE_CTRLB_DETACH;
+}
+
+static void usb_detach(sam0_common_usb_t *dev)
+{
+    DEBUG("sam_usb: detaching to host\n");
+    dev->config->device->CTRLB.reg |= USB_DEVICE_CTRLB_DETACH;
+}
+
+static int _usbdev_get(usbdev_t *usbdev, usbopt_t opt,
+                       void *value, size_t max_len)
+{
+    (void)usbdev;
+    (void)max_len;
+    int res = -ENOTSUP;
+    switch (opt) {
+        case USBOPT_MAX_VERSION:
+            assert(max_len == sizeof(usb_version_t));
+            *(usb_version_t *)value = USB_VERSION_20;
+            res = sizeof(usb_version_t);
+            break;
+        case USBOPT_MAX_SPEED:
+            assert(max_len == sizeof(usb_speed_t));
+            *(usb_speed_t *)value = USB_SPEED_FULL;
+            res = sizeof(usb_speed_t);
+            break;
+        default:
+            DEBUG("Unhandled get call: 0x%x\n", opt);
+            break;
+    }
+    return res;
+}
+
+static int _usbdev_set(usbdev_t *dev, usbopt_t opt,
+                       const void *value, size_t value_len)
+{
+    sam0_common_usb_t *usbdev = (sam0_common_usb_t *)dev;
+
+    (void)value_len;
+    int res = -ENOTSUP;
+    switch (opt) {
+        case USBOPT_ADDRESS:
+            assert(value_len == sizeof(uint8_t));
+            uint8_t addr = (*((uint8_t *)value));
+            _set_address(usbdev, addr);
+            break;
+        case USBOPT_ATTACH:
+            assert(value_len == sizeof(usbopt_enable_t));
+            if (*((usbopt_enable_t *)value)) {
+                usb_attach(usbdev);
+            }
+            else {
+                usb_detach(usbdev);
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        default:
+            DEBUG("Unhandled set call: 0x%x\n", opt);
+            break;
+    }
+    return res;
+}
+
+static void _ep_disable(usbdev_ep_t *ep)
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+
+    if (ep->dir == USB_EP_DIR_OUT) {
+        ep_reg->EPCFG.bit.EPTYPE0 = 0;
+    }
+    else {
+        ep_reg->EPCFG.bit.EPTYPE1 = 0;
+    }
+}
+
+static void _ep_enable(usbdev_ep_t *ep)
+{
+    DEBUG("sam_usb: Enabling endpoint %d, dir %s\n", ep->num,
+          ep->dir == USB_EP_DIR_OUT ? "OUT" : "IN");
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+    uint8_t type = 0;
+
+    switch (ep->type) {
+        case USB_EP_TYPE_CONTROL:
+            type = 0x01;
+            break;
+        case USB_EP_TYPE_ISOCHRONOUS:
+            type = 0x02;
+            break;
+        case USB_EP_TYPE_BULK:
+            type = 0x03;
+            break;
+        case USB_EP_TYPE_INTERRUPT:
+            type = 0x04;
+            break;
+        case USB_EP_TYPE_NONE:
```

Comment:
add default statement at the same time here ?

Reply:
How about I add an
```
default:
  assert(false);
```
here. Anything else than the types already handled in the switch is an error.

Reply:
Never mind, I can just change the `case USB_EP_TYPE_NONE:` to `default:`

Reply:
See 279e8449cbfb441ceda9f7821ef0ca4a13319a60

Reply:
That would be perfect.